src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/PackagePropertyPage/PackageIconValueProvider.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Properties.Package
+{
+    [ExportInterceptingPropertyValueProvider(PackageIconPropertyName, ExportInterceptingPropertyValueProviderFile.ProjectFile)]
+    internal sealed class PackageIconValueProvider : InterceptingPropertyValueProviderBase
+    {
+        private const string PackageIconPropertyName = "PackageIcon";
+        private const string PackPropertyName = "Pack";
+        private const string PackagePathPropertyName = "PackagePath";
+        private const string NoneItem = "None";
+        private const string TrueValue = "True";
+
+        private readonly IProjectItemProvider _sourceItemsProvider;
+        private readonly UnconfiguredProject _unconfiguredProject;
+
+        [ImportingConstructor]
+        public PackageIconValueProvider(
+            [Import(ExportContractNames.ProjectItemProviders.SourceFiles)] IProjectItemProvider sourceItemsProvider,
+            UnconfiguredProject unconfiguredProject)
+        {
+            _sourceItemsProvider = sourceItemsProvider;
+            _unconfiguredProject = unconfiguredProject;
+        }
+
+        // https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets#packageicon
+        private static string CreatePackageIcon(string filePath, string? packagePath)
+        {
+            string filename = Path.GetFileName(filePath);
+            // Make a slash-only value into empty string, so it won't get prepended onto the path.
+            packagePath = ("\\".Equals(packagePath, StringComparison.Ordinal) || "/".Equals(packagePath, StringComparison.Ordinal))
+                ? string.Empty : (packagePath ?? string.Empty);
+            // The assumption is that PackagePath does not contain a path to a file; only a directory path.
+            return Path.Combine(packagePath, filename);
+        }
+
+        private async Task<IProjectItem?> GetExistingNoneItemAsync(string existingPackageIcon)
+        {
+            foreach (IProjectItem noneItem in await _sourceItemsProvider.GetItemsAsync(NoneItem))
```

Comment:
You may want to test and see what happens if the matching `None` item is found in a Directory.Build.props or other imported file.

Reply:
This was a great suggestion. I tested this an the results were... bizarre. The `SetUnevaluatedIncludeAsync` call doesn't function if the None item is not inside the project file. This would cause a desync between the in-memory representation of the project file and the one on-disk. You'd even see the Solution Explorer updating the included files, but there would be no new None item in the project file. After a couple iterations, I got the functionality working as it should.

Basically, I still need to locate the None item, as logic related to `PackagePath` uses that information. So, I only check if the existing None item is within the project file right before either updating the existing item or creating a new one. This way, the PackagePath information is preserved. Now, you might say, "Well, if I add the file that already exists in a `.props` file, what happens?" This is the magic. It will call `SetUnevaluatedIncludeAsync` and set it to the same Include path as that in the `.props` file's None item. Then, deduplication occurs on the `.csproj`, thus removing the None item in the `.csproj`! It is really clean, honestly. I've made a GIF to show you what I mean. So, issues related to this are now fixed.

![PackageIconExternalFileExample](https://user-images.githubusercontent.com/17788297/127241911-9dff5cfc-f4c3-43dc-8722-5be4bdbad893.gif)